kapitan/refs/secrets/vault.py

```diff
@@ -0,0 +1,167 @@
+    * VAULT_AUTHTYPE: authentication type to use: token, userpass, github, ldap, approle
+    * VAULT_TOKEN: token for vault
+    * VAULT_ROLE_ID: (required by approle)
+    * VAULT_SECRET_ID: (required by approle)
+    * VAULT_USER: username to login to vault
+    * VAULT_PASSWORD: password to login to vault
+    * VAULT_CLIENT_KEY: path to an unencrypted PEM-encoded private key matching the client certificate
+    * VAULT_CLIENT_CERT: path to a PEM-encoded client certificate for TLS authentication to the Vault server
+    * VAULT_CACERT: path to a PEM-encoded CA cert file to use to verify the Vault server TLS certificate
+    * VAULT_CAPATH: path to a directory of PEM-encoded CA cert files to verify the Vault server TLS certificate
+    * VAULT_NAMESPACE: specify the Vault Namespace, if you have one
+    """
+    env = {}
+    env['url'] = getenv( 'VAULT_ADDR', default='http://127.0.0.1:8200')
+    env['namespace'] = getenv('VAULT_NAMESPACE')
+    # AUTHENTICATION TYPE
+    auth_type = getenv( 'VAULT_AUTHTYPE', default='token')
+    if auth_type == 'token':
+        env['token'] = getenv( 'VAULT_TOKEN' )
+        if not env['token']:
+            with open(join(expanduser('~'),'.vault-token'),'r') as f:
+                env['token'] = f.read()
+    elif auth_type == 'userpass':
+        env['username'] = getenv( 'VAULT_USER' )
+        env['password'] = getenv( 'VAULT_PASSWORD' )
+    elif auth_type == 'approle':
+        env['role_id'] = getenv('VAULT_ROLE_ID')
+        env['secret_id'] = getenv( 'VAULT_SECRET_ID' )
+    # VERIFY VAULT SERVER TLS CERTIFICATE
+    verify = getenv( 'VAULT_SKIP_VERIFY', default='')
+    if verify.lower() == 'true':
+        env['verify'] = False
+    elif verify.lower() == 'false':
+        cert = getenv( 'VAULT_CACERT' )
+        if not cert:
+            cert_path = getenv( 'VAULT_CAPATH' )
+            if not cert_path:
+                raise Exception('Neither VAULT_CACERT nor VAULT_CAPATH specified')
+            env['verify'] = cert_path
+        else:
+            env['verify'] = cert
+    # CLIENT CERTIFICATE FOR TLS AUTHENTICATION
+    client_key,client_cert = getenv( 'VAULT_CLIENT_KEY' ), getenv( 'VAULT_CLIENT_CERT' ) 
+    if client_key != None and client_cert != None:
+        env['cert'] = (client_cert,client_key)
+    return env
+
+def vault_obj():
+    client = hvac.Client(
+        **{k:v for k,v in get_env().items() if v is not None}
+    )
+    assert (
+        client.is_authenticated()
+    ), "Vault Authentication Error, Environment Variables defined?"
+    return client
+
+
+class VaultSecret(Ref):
+
+    """
+    Hashicorp Vault can be used if using KV Secret Engine
+    """
+
+    def __init__(self,data,encode_base64=False,**kwargs):
+        """
+        set encoding_base64 to True to base64 encoding key before encrypting and writing
+        """
+        self._encrypt(data, encode_base64)
```

Comment:
This would include a large number, would request @uberspot @ramaro to suggest the one they think should be included or are planning to include in the future.